pybmix/core/pybmixcpp/py_hier/python_hierarchy.h

```diff
@@ -0,0 +1,369 @@
+  downcast_state(google::protobuf::Message *const state_) const {
+    return google::protobuf::internal::down_cast<
+        bayesmix::AlgorithmState::ClusterState *>(state_);
+  };
+
+  //! Down-casts the given generic proto message to a ClusterState proto
+  const bayesmix::AlgorithmState::ClusterState &
+  downcast_state(const google::protobuf::Message &state_) const {
+    return google::protobuf::internal::down_cast<
+        const bayesmix::AlgorithmState::ClusterState &>(state_);
+  };
+
+  //! Down-casts the given generic proto message to a HierarchyHypers proto
+  bayesmix::AlgorithmState::HierarchyHypers *
+  downcast_hypers(google::protobuf::Message *const state_) const {
+    return google::protobuf::internal::down_cast<
+        bayesmix::AlgorithmState::HierarchyHypers *>(state_);
+  };
+
+  //! Down-casts the given generic proto message to a HierarchyHypers proto
+  const bayesmix::AlgorithmState::HierarchyHypers &
+  downcast_hypers(const google::protobuf::Message &state_) const {
+    return google::protobuf::internal::down_cast<
+        const bayesmix::AlgorithmState::HierarchyHypers &>(state_);
+  };
+
+  //! Container for state values
+  Python::State state;
+
+  //! Container for prior hyperparameters values
+  std::shared_ptr<Python::Hyperparams> hypers;
+
+  //! Container for posterior hyperparameters values
+  Python::Hyperparams posterior_hypers;
+
+  //! Pointer to a Protobuf prior object for this class
+  std::shared_ptr<bayesmix::PythonHierPrior> prior;
+
+  //! Set of indexes of data points belonging to this cluster
+  std::set<int> cluster_data_idx;
+
+  //! Current cardinality of this cluster
+  int card = 0;
+
+  //! Logarithm of current cardinality of this cluster
+  double log_card = stan::math::NEGATIVE_INFTY;
+
+  //! Pointer to the dataset matrix for the mixture model
+  const Eigen::MatrixXd *dataset_ptr = nullptr;
+
+  //! Set of values of data points belonging to this cluster
+  // std::list<Eigen::RowVectorXd> cluster_data_values;
+  Eigen::MatrixXd cluster_data_values;
+
+  //! Vector of summary statistics
+  std::vector<double> sum_stats;
+
+  //! Py objects for the rng
+  py::module_ numpy = py::module_::import("numpy");
+  py::module_ numpy_random = py::module_::import("numpy.random");
+  py::object py_engine = numpy_random.attr("MT19937")();
+  py::object py_gen = numpy_random.attr("Generator")(py_engine);
+
+  //! Py module where the hierarchy is implemented
+  py::module_ hier_implementation;
+
+  //! Py methods implemented in hier_implementation
+  py::object clear_summary_statistics_evaluator;
+  py::object draw_evaluator;
+  py::object initialize_state_evaluator;
+  py::object initialize_hypers_evaluator;
+  py::object is_conjugate_evaluator;
+  py::object like_lpdf_evaluator;
+  py::object marg_lpdf_evaluator;
+  py::object posterior_hypers_evaluator;
+  py::object sample_full_cond_evaluator;
+  py::object update_summary_statistics_evaluator;
+  // py::object update_hypers_evaluator;
```

Comment:
why is this commented?

Reply:
It was not implemented, we have just added it.

Reply:
Now `NNIG_Hierarchy_1.py` implements the NNIG hierarchy with fixed priors (no prior, just fixed values) and `NNIG_Hierarchy_2.py` implements the NNIG hierarchy with NGGPrior (prior on the mean, var_scaling, and scale).